src/server/pfs/server/source.go

```diff
@@ -51,10 +53,16 @@ func NewSource(commitInfo *pfs.CommitInfo, fs fileset.FileSet, opts ...SourceOpt
 		s.dirIndexOpts = append(s.dirIndexOpts, index.WithRange(&index.PathRange{
 			Lower: sc.pathRange.Lower,
 		}))
+		// The upper for the index path range is set to be past the provided upper
+		// to ensure that directories between the last file that should be emitted
+		// and the first file that shouldn't get created.
+		// For example, the files /d1/f1 and /d2/f2 with a path range of [/d1/f1, /d2/f2) should
+		// emit /d1/f1 and /d2/.
 		s.fileIndexOpts = append(s.fileIndexOpts, index.WithRange(&index.PathRange{
 			Lower: sc.pathRange.Lower,
-			Upper: sc.pathRange.Upper,
+			Upper: sc.pathRange.Upper + "_",
```

Comment:
Is there any value to using `"\0"` here?

Reply:
It doesn't affect the results, but it is probably clearer what is being done if the null byte is used.